pep-0544.txt

```diff
@@ -0,0 +1,967 @@
+
+  class Example(Protocol):
+      def first(self) -> int:     # This is a protocol member
+          return 42
+
+      @abstractmethod
+      def second(self) -> int:    # Method without a default implementation
+          """Some method."""
+
+Note that although formally the implicit return type of a method with
+a trivial body is ``None``, type checker will not warn about above example,
+such convention is similar to how methods are defined in stub files.
+Static methods, class methods, and properties are equally allowed
+in protocols.
+
+To define a protocol variable, one must use PEP 526 variable
+annotations in the class body. Additional attributes *only* defined in
+the body of a method by assignment via ``self`` are not allowed. The rationale
+for this is that the protocol class implementation is often not shared by
+subtypes, so the interface should not depend on the default implementation.
+Examples::
+
+  from typing import Protocol, List
+
+  class Template(Protocol):
+      name: str        # This is a protocol member
+      value: int = 0   # This one too (with default)
+
+      def method(self) -> None:
+          self.temp: List[int] = [] # Error in type checker
+
+To distinguish between protocol class variables and protocol instance
+variables, the special ``ClassVar`` annotation should be used as specified
+by PEP 526.
+
+
+Explicitly declaring implementation
+-----------------------------------
+
+To explicitly declare that a certain class implements the given protocols,
+they can be used as regular base classes. In this case a class could use
+default implementations of protocol members. ``typing.Sequence`` is a good
+example of a protocol with useful default methods.
+
+Abstract methods with trivial bodies are recognized by type checkers as
+having no default implementation and can't be used via ``super()`` in
+explicit subclasses. The default implementations can not be used if
+the subtype relationship is implicit and only via structural
+subtyping -- the semantics of inheritance is not changed. Examples::
+
+    class PColor(Protocol):
+        @abstractmethod
+        def draw(self) -> str:
+            ...
+        def complex_method(self) -> int:
+            # some complex code here
+
+    class NiceColor(PColor):
+        def draw(self) -> str:
+            return "deep blue"
+
+    class BadColor(PColor):
+        def draw(self) -> str:
+            return super().draw()  # Error, no default implementation
+
+    class ImplicitColor:   # Note no 'PColor' base here
+        def draw(self) -> str:
+            return "probably gray"
+        def comlex_method(self) -> int:
+            # class needs to implement this
+
+    nice: NiceColor
+    another: ImplicitColor
+
+    def represent(c: PColor) -> None:
+        print(c.draw(), c.complex_method())
+
+    represent(nice) # OK
+    represent(another) # Also OK
+
+Note that there is no conceptual difference between explicit and implicit
```

Comment:
Well, hm, there seem to be plenty of subtle differences, e.g. the requirement to implement everything only in an implicit subclass.